### PR TITLE
Global Firewall remediation

### PR DESCRIPTION
**Pull request by dgwhited**
Remediates firewalls with 0.0.0.0/0 source ranges when created or modified.

Also adds a feature to set function memory independently because this function runs over 128mb after garbage collection and other trimming efforts.

### CODE REVIEW

File: src/public_firewall_port/main.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        logging.info(f'The project {project_id} is not in the allowlist, is in the denylist, or a list is not fully configured. Continuing evaluation.')
+        check_resource(project_id, firewall_name)
+    else:
+        logging.info(f'The project {project_id} is in the allowlist or is not in the denylist. No action being taken.')
+
+    # This function sits around 133mb of memory usage at creation date, if we don't cleanup, high activity projects may run into issues.
+    del log_entry
+    gc.collect()
+
+def check_resource(project_id, firewall_name):
```
Comment (jdyke) [inline src/public_firewall_port/main.py:41]: Can you add a docstring to this function?

File: src/public_firewall_port/main.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        if mode == "read":
+            logging.info('Lockdown is in read-only mode. Taking no action.')
+    else:
+        logging.info(f"The firewall rule: {firewall_name} is not enabled or 0.0.0.0/0 was removed.")
+
+    # More cleanup
+    del compute_client
+
+def describe_firewall(compute_client, firewall_name, project_id):
+    """ Gets information about the firewall
```
Comment (jdyke) [inline src/public_firewall_port/main.py:92]: Should this description be inside the docstring?

File: src/public_firewall_port/main.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    try:
+        request = compute_client.firewalls().get(project=project_id, firewall=firewall_name)
+        response = request.execute(num_retries=5)
+    except:
+        logging.info(f'Could not retrieve enablement status for {firewall_name}.')
```
Comment (jdyke) [inline src/public_firewall_port/main.py:98]: `logging.error` vs `logging.info` ?

File: src/public_firewall_port/main.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        raise
+    return response
+
+def disable_firewall(compute_client, firewall_name, project_id):
+    """ Disables a firewall rule
```
Comment (jdyke) [inline src/public_firewall_port/main.py:103]: ditto above description inside docstring

File: src/public_firewall_port/main.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    firewall_body = {
+    "name": firewall_name,
+    "disabled": "true"
+    }
+    try:
+        request = compute_client.firewalls().patch(project=project_id, firewall=firewall_name, body=firewall_body)
+        response = request.execute(num_retries=5)
+    except:
+        logging.info(f'Could not disable firewall: {firewall_name}.')
```
Comment (jdyke) [inline src/public_firewall_port/main.py:113]: `logging.error` vs `logging.info` ?

**Review by jdyke — CHANGES_REQUESTED** [review]

File: src/public_firewall_port/main.py (hunk 3 above)
Comment (dgwhited) [reply src/public_firewall_port/main.py]: good call

**Review by jdyke — APPROVED** [review]
LGTM, thanks for contributing this use case @dgwhited

**Pull request merged**